pkg/logcli/query/query.go

```diff
@@ -72,26 +73,91 @@ func (q *Query) DoQuery(c *client.Client, out output.LogOutput, statistics bool)
 
 	if q.isInstant() {
 		resp, err = c.Query(q.QueryString, q.Limit, q.Start, d, q.Quiet)
+		if err != nil {
+			log.Fatalf("Query failed: %+v", err)
+		}
+		if statistics {
+			q.printStats(resp.Data.Statistics)
+		}
+		_, _ = q.printResult(resp.Data.Result, out, nil)
 	} else {
-		resp, err = c.QueryRange(q.QueryString, q.Limit, q.Start, q.End, d, q.Step, q.Interval, q.Quiet)
-	}
-
-	if err != nil {
-		log.Fatalf("Query failed: %+v", err)
-	}
+		if q.Limit < q.BatchSize {
+			q.BatchSize = q.Limit
+		}
+		resultLength := q.BatchSize
+		total := 0
+		start := q.Start
+		end := q.End
+		var lastEntry []*loghttp.Entry
+		for total < q.Limit {
+			bs := q.BatchSize
+			// We want to truncate the batch size if the remaining number
+			// of items needed to reach the limit is less than the batch size
+			if q.Limit-total < q.BatchSize {
+				// Truncated batchsize is q.Limit - total, however we add to this
+				// the length of the overlap from the last query to make sure we get the
+				// correct amount of new logs knowing there will be some overlapping logs returned.
+				bs = q.Limit - total + len(lastEntry)
+			}
+			resp, err = c.QueryRange(q.QueryString, bs, start, end, d, q.Step, q.Interval, q.Quiet)
+			if err != nil {
+				log.Fatalf("Query failed: %+v", err)
+			}
+
+			if statistics {
+				q.printStats(resp.Data.Statistics)
+			}
+
+			resultLength, lastEntry = q.printResult(resp.Data.Result, out, lastEntry)
+			// Was not a log stream query, or no results, no more batching
+			if resultLength <= 0 {
+				break
+			}
+			// Also no result, wouldn't expect to hit this.
+			if lastEntry == nil || len(lastEntry) == 0 {
+				break
+			}
+			// Can only happen if all the results return in one request
+			if resultLength == q.Limit {
+				break
+			}
+			if len(lastEntry) >= q.BatchSize {
+				log.Fatalf("Invalid batch size %v, the next query will have %v overlapping entries "+
```

Comment:
I think it'd be better to just return up to `limit`, even if it doesn't consume all entries with the same timestamp. Not a blocker though.

Reply:
Batches are printed as they are received so in this case we would print the last batch but then fail before going after the next (because it would effectively have exactly the same entries as the last batch)

Reply:
What this error is saying is basically you queries with a batch size of 1000 and all 1000 entries have the same timestamp so you have to increase your batch size until that isn't true.

With a batch size default of 1000 I hope it's pretty unlikely people will ever run into this.

Reply:
I think whoever finds themselves hitting this error should open a pull request to make this better :)